### PR TITLE
Warrior specialization features and stances

### PR DESCRIPTION
**Pull request by Swillis57**
Includes masteries and spec passives, talents and spells will be done in separate PRs to keep them small.

### CODE REVIEW

File: sim/core/buffs.go [hunk 1]
```diff
@@ -772,23 +772,29 @@ func registerUnholyFrenzyCD(agent Agent, numUnholyFrenzy int32) {
 		numUnholyFrenzy)
 }
 
-func UnholyFrenzyAura(character *Unit, actionTag int32) *Aura {
+func UnholyFrenzyAura(agent Agent, actionTag int32) *Aura {
```
Comment (rosenrusinov) [inline sim/core/buffs.go:775]: Dont use Agent here and instead use Character if you dont need the agent

File: sim/core/utils.go [hunk 2]
```diff
@@ -158,6 +158,11 @@ func Flatten[T any](src [][]T) []T {
 	return dst
 }
 
+func MasteryRatingToMasteryPoints(masteryRating float64) float64 {
+	// TODO: check mastery point rounding behavior
+	return math.Floor(masteryRating / MasteryRatingPerMasteryPoint)
```
Comment (rosenrusinov) [inline sim/core/utils.go:163]: Is this floor supposed to be here? If that is true then a character gaining 178 mastery rating could mean they get nothing because of rounding and I highly doubt the system works like that. This is close to a whole item worth of mastery being ignored

Comment (rosenrusinov) [reply sim/core/utils.go:163]: I would suggest you remove the flooring for now as we've seen numerous times that stats are not rounded in any way. (Even the expertise in wotlk was never rounded/truncated as people used to believe)

File: sim/death_knight/_unholy_frenzy.go [hunk 3]
```diff
@@ -20,7 +20,13 @@ func (dk *DeathKnight) registerUnholyFrenzyCD() {
 	if unholyFrenzyTarget == nil {
 		return
 	}
-	dk.UnholyFrenzyAura = core.UnholyFrenzyAura(unholyFrenzyTarget, actionID.Tag)
+
+	unholyFrenzyAgent := dk.Env.GetAgentFromUnit(unholyFrenzyTarget)
+	if unholyFrenzyAgent == nil {
+		return
+	}
```
Comment (rosenrusinov) [inline sim/death_knight/_unholy_frenzy.go:27]: revert unholy frenzy to use character and not agent and revert this change

File: sim/warrior/arms/arms.go [hunk 4]
```diff
@@ -39,28 +42,95 @@ func NewArmsWarrior(character *core.Character, options *proto.Player) *ArmsWarri
...
+	rbo := core.RageBarOptions{
+		StartingRage: armsOptions.ClassOptions.StartingRage,
+		OnHitDealtRageGain: func(sim *core.Simulation, spell *core.Spell, result *core.SpellResult, rage float64) float64 {
+			return rage * 1.25 // Endless Rage is now part of Anger Management, now an Arms specialization passive
+		},
+	}
+	if mh := war.GetMHWeapon(); mh != nil {
+		rbo.MHSwingSpeed = mh.SwingSpeed
+	}
+	war.EnableRageBar(rbo)
```
Comment (rosenrusinov) [inline sim/warrior/arms/arms.go:54]: This multiplier can just be passed down as a multiplier to the rage bar instead of having to do a complete new callback every time you do a hit. It seems like an overkill for just a static 25% increased rage gen on hits.

File: sim/warrior/arms/arms.go [hunk 5]
```diff
@@ -39,28 +42,95 @@ func NewArmsWarrior(character *core.Character, options *proto.Player) *ArmsWarri
...
-	// war.EnableAutoAttacks(war, core.AutoAttackOptions{
-	// 	MainHand:       war.WeaponFromMainHand(war.DefaultMeleeCritMultiplier()),
-	// 	OffHand:        war.WeaponFromOffHand(war.DefaultMeleeCritMultiplier()),
-	// 	AutoSwingMelee: true,
-	// 	ReplaceMHSwing: war.TryHSOrCleave,
-	// })
+	war.EnableAutoAttacks(war, core.AutoAttackOptions{
+		MainHand:       war.WeaponFromMainHand(war.DefaultMeleeCritMultiplier()),
+		AutoSwingMelee: true,
+	})
```
Comment (rosenrusinov) [inline sim/warrior/arms/arms.go:59]: Can't arms warrior dual wield anymore? The way you've changed this you have removed support for dual wielding completely

File: sim/warrior/arms/arms.go [hunk 6]
```diff
@@ -39,28 +42,95 @@ func NewArmsWarrior(character *core.Character, options *proto.Player) *ArmsWarri
...
+
+func (war *ArmsWarrior) RegisterMastery() {
+	// TODO:
+	//	test what things the extra attack can proc
+	//	does the extra attack use the same hit table
+	//  can it proc off of missed/dodged/parried attacks
+	//
+	// 4.3.3 simcraft implements SoO as a standard autoattack with a 0.5s ICD
+	procAttackConfig := *war.AutoAttacks.MHConfig()
+	procAttackConfig.ActionID = core.ActionID{SpellID: StrikesOfOpportunityHitID, Tag: procAttackConfig.ActionID.Tag}
```
Comment (rosenrusinov) [inline sim/warrior/arms/arms.go:102]: Are you not overwriting the ActionID of the auto attacks main hand hit by changing its config here? Or is this gonna create a duplicate instead?

File: sim/warrior/arms/arms.go [hunk 7]
```diff
@@ -39,28 +42,95 @@ func NewArmsWarrior(character *core.Character, options *proto.Player) *ArmsWarri
...
+		Timer:    war.NewTimer(),
+		Duration: time.Millisecond * 500,
+	}
+
+	war.RegisterAura(core.Aura{
+		Label:    "Strikes of Opportunity",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
```
Comment (rosenrusinov) [inline sim/warrior/arms/arms.go:115]: Instead of adding these 4 lines you can do `core.MakePermanent(war.RegisterAura(...))`

File: sim/warrior/fury/fury.go [hunk 8]
```diff
@@ -39,28 +40,52 @@ func NewFuryWarrior(character *core.Character, options *proto.Player) *FuryWarri
...
+
+func (war *FuryWarrior) RegisterMastery() {
+	// The actual effects of Unshackled Fury need to be handled by specific spells
+	// as it modifies the "benefit" of them (e.g. it both increases Raging Blow's damage
+	// and Enrage's damage bonus) so we just keep this field up to date for spells to read
+	war.PseudoStats.EnragedEffectMultiplier = war.GetMasteryBonusMultiplier()
+	war.AddOnMasteryStatChanged(func(sim *core.Simulation, oldMastery, newMastery float64) {
+		war.PseudoStats.EnragedEffectMultiplier = war.GetMasteryBonusMultiplier()
+	})
+}
```
Comment (rosenrusinov) [inline sim/warrior/fury/fury.go:87]: If this mastery affects specific spells then there is no reason to keep this EnragedEffectMultiplier as a cached stat. The spell can just calculate the effect when they are being executed and you will get rid of a very warrior specific stat in the pseudoStats which to me is kinda ugly

File: sim/warrior/protection/protection.go [hunk 9]
```diff
@@ -37,35 +40,88 @@ func NewProtectionWarrior(character *core.Character, options *proto.Player) *Pro
...
+	}
+	if mh := war.GetMHWeapon(); mh != nil {
+		rbo.MHSwingSpeed = mh.SwingSpeed
+	}
+
+	war.EnableRageBar(rbo)
+	war.EnableAutoAttacks(war, core.AutoAttackOptions{
+		MainHand:       war.WeaponFromMainHand(war.DefaultMeleeCritMultiplier()),
+		AutoSwingMelee: true,
+	})
```
Comment (rosenrusinov) [inline sim/warrior/protection/protection.go:62]: no support for dual wielding as a protection warrior in cata?

File: sim/warrior/protection/protection.go [hunk 10]
```diff
@@ -37,35 +40,88 @@ func NewProtectionWarrior(character *core.Character, options *proto.Player) *Pro
...
+
+	war.RegisterSpecializationEffects()
```
Comment (rosenrusinov) [inline sim/warrior/protection/protection.go:64]: This should be called from the specs Initialize() as you dont have all stats created during the spec struct creation.

File: sim/warrior/stances.go [hunk 11]
```diff
@@ -62,41 +63,38 @@ func (warrior *Warrior) makeStanceSpell(stance Stance, aura *core.Aura, stanceCD
...
 	actionID := core.ActionID{SpellID: 71}
-	if warrior.Talents.ImprovedDefensiveStance > 0 {
+	if warrior.Talents.BastionOfDefense > 0 {
 		enrageAura := warrior.GetOrRegisterAura(core.Aura{
 			Label:    "Enrage",
 			ActionID: core.ActionID{SpellID: 57516},
 			Duration: 12 * time.Second,
 			OnGain: func(aura *core.Aura, sim *core.Simulation) {
-				aura.Unit.PseudoStats.SchoolDamageDealtMultiplier[stats.SchoolIndexPhysical] *= 1.0 + 0.05*float64(warrior.Talents.ImprovedDefensiveStance)
+				aura.Unit.PseudoStats.SchoolDamageDealtMultiplier[stats.SchoolIndexPhysical] *= 1.0 + 0.05*float64(warrior.Talents.BastionOfDefense)
```
Comment (rosenrusinov) [inline sim/warrior/stances.go:94]: Cache the value outside of the OnGain/OnExpire effects

File: sim/warrior/stances.go [hunk 12]
```diff
@@ -105,86 +103,40 @@ func (warrior *Warrior) registerDefensiveStanceAura() {
 			Duration: core.NeverExpires,
 			OnSpellHitTaken: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
 				if result.Outcome.Matches(core.OutcomeBlock | core.OutcomeDodge | core.OutcomeParry) {
-					if sim.RandomFloat("Enrage Trigger Chance") <= 0.5*float64(warrior.Talents.ImprovedDefensiveStance) {
+					if sim.RandomFloat("Enrage Trigger Chance") <= 0.1*float64(warrior.Talents.BastionOfDefense) {
```
Comment (rosenrusinov) [inline sim/warrior/stances.go:106]: You should cache any values that dont depend on dynamic values outside of the callbacks.

File: sim/warrior/arms/arms.go (hunk 5 above)
Comment (Swillis57) [reply sim/warrior/arms/arms.go:59]: Only Fury is able to dual wield now. See the wording of https://www.wowhead.com/cata/spell=23588/dual-wield-specialization

There's also a patch note about it here https://wowpedia.fandom.com/wiki/Dual_wield#Patch_changes

File: sim/warrior/arms/arms.go (hunk 6 above)
Comment (Swillis57) [reply sim/warrior/arms/arms.go:102]: This copies the MH hit spell and changes the ActionID to distinguish it. From what I can find the proc appears to just be an additional MH swing with the same attributes.

File: sim/warrior/protection/protection.go (hunk 9 above)
Comment (Swillis57) [reply sim/warrior/protection/protection.go:62]: See https://github.com/wowsims/cata/pull/33#discussion_r1535978548

File: sim/core/buffs.go [hunk 13]
```diff
@@ -772,23 +772,29 @@ func registerUnholyFrenzyCD(agent Agent, numUnholyFrenzy int32) {
...
 	actionID := ActionID{SpellID: 49016, Tag: actionTag}
 
+	character := agent.GetCharacter()
+
+	// If this character has an effect that multiplies enrage effects, incorporate it
+	// Mainly for Fury Warrior's Unshackled Rage mastery
+	effect := 1.0 + (0.2 * character.PseudoStats.EnragedEffectMultiplier)
```
[the remark this answers is not in the document]
Comment (Swillis57) [reply sim/core/buffs.go:782]: SimC implemented this interaction but it's not clear looking at the mastery's spell effects if it would or not. I'll revert this and leave a note to test it

**Review by rosenrusinov — APPROVED** [review]

File: sim/warrior/arms/arms.go (hunk 5 above)
Comment (rosenrusinov) [reply sim/warrior/arms/arms.go:57]: Interesting. Did not know they flat out removed it. Leaving a note to self to limit the availability of equipping OH weapons in those specs that no longer have DW

**Review by rosenrusinov — APPROVED** [review]
